les/fetcher.go

```diff
@@ -23,172 +23,410 @@ import (
+// canonizing new heads. It also helps to always download the minimum necessary
+// amount of headers with a single request.
+type fetcherTreeNode struct {
+	hash             common.Hash
+	number           uint64
+	td               *big.Int
+	known, requested bool
+	parent           *fetcherTreeNode
+	children         []*fetcherTreeNode
+}
+
+// fetchRequest represents a header download request
 type fetchRequest struct {
-	hash   common.Hash
-	amount uint64
-	peer   *peer
+	hash    common.Hash
+	amount  uint64
+	peer    *peer
+	sent    mclock.AbsTime
+	timeout bool
 }
 
+// fetchResponse represents a header download response
 type fetchResponse struct {
 	reqID   uint64
 	headers []*types.Header
+	peer    *peer
 }
 
+// newLightFetcher creates a new light fetcher
 func newLightFetcher(pm *ProtocolManager) *lightFetcher {
 	f := &lightFetcher{
-		pm:              pm,
-		chain:           pm.blockchain,
-		odr:             pm.odr,
-		headAnnouncedBy: make(map[common.Hash][]*peer),
-		deliverChn:      make(chan fetchResponse, 100),
-		requested:       make(map[uint64]fetchRequest),
-		timeoutChn:      make(chan uint64),
-		notifyChn:       make(chan bool, 100),
-		syncDone:        make(chan struct{}),
-		currentTd:       big.NewInt(0),
+		pm:             pm,
+		chain:          pm.blockchain.(*light.LightChain),
+		odr:            pm.odr,
+		peers:          make(map[*peer]*fetcherPeerInfo),
+		deliverChn:     make(chan fetchResponse, 100),
+		requested:      make(map[uint64]fetchRequest),
+		timeoutChn:     make(chan uint64),
+		requestChn:     make(chan bool, 100),
+		syncDone:       make(chan *peer),
+		maxConfirmedTd: big.NewInt(0),
 	}
 	go f.syncLoop()
 	return f
 }
 
-func (f *lightFetcher) notify(p *peer, head *announceData) {
-	var headHash common.Hash
-	if head == nil {
-		// initial notify
-		headHash = p.Head()
-	} else {
-		if core.GetTd(f.pm.chainDb, head.Hash, head.Number) != nil {
-			head.haveHeaders = head.Number
-		}
-		//fmt.Println("notify", p.id, head.Number, head.ReorgDepth, head.haveHeaders)
-		if !p.addNotify(head) {
-			//fmt.Println("addNotify fail")
-			f.pm.removePeer(p.id)
+// syncLoop is the main event loop of the light fetcher
+func (f *lightFetcher) syncLoop() {
+	f.pm.wg.Add(1)
+	defer f.pm.wg.Done()
+
+	requestStarted := false
+	for {
+		select {
+		case <-f.pm.quitSync:
+			return
+		// when a new announce is received, request loop keeps running until
+		// no further requests are necessary or possible
+		case newAnnounce := <-f.requestChn:
+			f.lock.Lock()
+			s := requestStarted
+			requestStarted = false
+			if !f.syncing && !(newAnnounce && s) {
+				if p, r, a := f.nextRequest(); r != nil {
+					requestStarted = true
+					reqID, started := f.request(p, r, a)
+					if started {
+						go func() {
+							time.Sleep(softRequestTimeout)
+							f.reqMu.Lock()
+							req, ok := f.requested[reqID]
+							if ok {
+								req.timeout = true
+								f.requested[reqID] = req
+							}
```

Comment:
Looks odd to me ... You just got `req` from `f.requested[reqID]`, and now you're assigning it back into `f.requested[reqID]` - I don't follow what's happening here.